pkg/controllers/networking/pod_ip_cache.go

```diff
@@ -0,0 +1,161 @@
+)
+
+type PodIPCache interface {
+	Record(podUID types.UID, podName, namespace string, ipInstanceNames []string)
+	Release(ipInstanceName, namespace string) error
+	Get(podName, namespace string) (bool, types.UID, []string)
+}
+
+type podAllocatedInfo struct {
+	podUID          types.UID
+	ipInstanceNames []string
+}
+
+type podIPCache struct {
+	*sync.RWMutex
+
+	// use "name/namespace" of ip instance as key
+	podToIP map[string]*podAllocatedInfo
+
+	// use "name/namespace" of ip instance as key and "name" of pod as value
+	ipToPod map[string]string
+}
+
+func NewPodIPCache(c client.Reader) (PodIPCache, error) {
+	cache := &podIPCache{
+		podToIP: map[string]*podAllocatedInfo{},
+		ipToPod: map[string]string{},
+		RWMutex: &sync.RWMutex{},
+	}
+
+	ipList, err := controllerutils.ListIPInstances(c)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ip := range ipList.Items {
+		podName := networkingv1.FetchBindingPodName(&ip)
+		if len(podName) != 0 {
+			var podUID types.UID
+
+			if len(ip.Spec.Binding.PodUID) != 0 {
+				podUID = ip.Spec.Binding.PodUID
+			} else if !networkingv1.IsReserved(&ip) {
+				// TODO: no longer need to get pod if all the ip instances is updated to the v1.2 version
+				pod, err := controllerutils.GetPod(c, podName, ip.Namespace)
+				if err != nil {
+					if err = client.IgnoreNotFound(err); err != nil {
+						return nil, fmt.Errorf("unable to get Pod %v for IPInstance %v: %v", podName, ip.Name, err)
+					}
+				}
+
+				if pod != nil {
+					podUID = pod.UID
+				}
+			}
+
+			var recordedIPInstances []string
+			if cache.podToIP[namespacedKey(podName, ip.Namespace)] == nil {
+				recordedIPInstances = nil
+			} else {
+				recordedIPInstances = cache.podToIP[namespacedKey(podName, ip.Namespace)].ipInstanceNames
+			}
+
+			// this is different from a normal Record action
+			cache.podToIP[namespacedKey(podName, ip.Namespace)] = &podAllocatedInfo{
+				podUID:          podUID,
+				ipInstanceNames: append(recordedIPInstances, ip.Name),
+			}
+
+			cache.ipToPod[namespacedKey(ip.Name, ip.Namespace)] = podName
+		}
+	}
+
+	return cache, nil
+}
+
+func (c *podIPCache) Record(podUID types.UID, podName, namespace string, ipInstanceNames []string) {
+	c.Lock()
+	defer c.Unlock()
+
+	// don't check if the pod exist, just overwrite it
+	c.podToIP[namespacedKey(podName, namespace)] = &podAllocatedInfo{
+		podUID:          podUID,
+		ipInstanceNames: ipInstanceNames,
+	}
+
+	for _, ipInstanceName := range ipInstanceNames {
+		c.ipToPod[namespacedKey(ipInstanceName, namespace)] = podName
+	}
+}
+
+func (c *podIPCache) Release(ipInstanceName, namespace string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	podName, exist := c.ipToPod[namespacedKey(ipInstanceName, namespace)]
+	if !exist {
+		return fmt.Errorf("ipToPod record not exist for ip instance %v", ipInstanceName)
+	}
+
+	delete(c.ipToPod, namespacedKey(ipInstanceName, namespace))
+
+	info := c.podToIP[namespacedKey(podName, namespace)]
```

Comment:
`info` here may not be found every time, maybe you should check this too.

Reply:
done